executor/bitmap.h

```diff
@@ -0,0 +1,57 @@
+// Copyright 2017 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/mman.h>
+
+typedef unsigned int uint32;
+typedef unsigned char byte;
+
+static uint32 pcstart;
+static uint32 pcsize;
+static uint32 pcend;
+
+inline byte* getBitmap()
+{
+	int f = open("/bitmap", O_RDONLY);
+	if (f < 0)
+		return NULL;
+	ssize_t ret = read(f, &pcstart, sizeof(uint32));
+	if (ret < 0)
+		return NULL;
+	ret = read(f, &pcsize, sizeof(uint32));
+	if (ret < 0)
+		return NULL;
+	// coverage filter will be disable
+	if (pcsize == 0)
+		return NULL;
+	pcend = pcstart + pcsize;
+	fprintf(stderr, "Coverage filter from %x to %x, size %x\n", pcstart, pcend, pcsize);
+
+	uint32 bitmapSize = (pcsize >> 4) / 8 + 1;
+	byte* bitmap = (byte*)mmap(NULL, bitmapSize + 2 * sizeof(uint32), PROT_READ, MAP_SHARED, f, 0);
+	if (bitmap == NULL)
+		return NULL;
+	bitmap += sizeof(uint32) * 2;
+	return bitmap;
+}
+
+inline bool coverage_filter(uint64 pc, byte* bitmap)
+{
+	// don't filter any pc
+	if (bitmap == NULL)
+		return true;
+	// prevent overflow while searching bitmap
+	uint32 pc32 = (uint32)(pc & 0xffffffff);
+	if (pc32 < pcstart || pc32 > pcend)
+		return false;
+	pc32 -= pcstart;
+	pc32 = pc32 >> 4;
```

Comment:
Hi @dvyukov , for cutting the bitmap into a smaller size, the lowest 4-bit will be dropped. Since the instruction length in different arches is different, is it appropriate here? Or, should we use full address bitmap?

Reply:
Did you use it on instruction level?
I would assume main uses will be on function level and functions a frequently aligned to 16 bytes. So this looks like a reasonable tradeoff.
But this logic needs a comment. I was confused by these shifts when first looked at the code.